### PR TITLE
Couch impl

### PR DESCRIPTION
**Pull request by leecindy**
Issue #350

### CODE REVIEW

**Comment by tsconn23** [comment]
@leecindy Is there a particular issue to which this PR applies? If so, please add it to your PR description. Also, build verification jobs are not passing. Please address.

**Comment by drasko** [comment]
This contribution looks good to me, but I would also like to understand use-case. Replacing MongoDB only in Export Services will not remove potential Mongo issues (if they exist) if it is used in other EdgeX components.

**Comment by feclare** [comment]
Tests are missing. It should be something similar to mongo-client_test.go

**Comment by tsconn23** [comment]
@leecindy Can you please look into the build issues? Can't pull this in until verify checks pass.

File: export/client/clients/couch-client.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	Compression string                   `json:"compression,omitempty"`
+	Enable      bool                     `json:"enable"`
+	Destination string                   `json:"destination,omitempty"`
+}
+
+// Return a pointer to the CouchClient
+func newCouchClient(config DBConfiguration) (*CouchClient, error) {
+	// Create the dial info for the Couch session
+
+	connectionString := "http://" + config.Username + ":" + config.Password + "@" + config.Host + ":" + strconv.Itoa(config.Port)
```
Comment (drasko) [inline export/client/clients/couch-client.go]: Please use `Sprintf()`

File: export/client/clients/couch-client.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+	timeout := time.Duration(config.Timeout) * time.Millisecond
+
+	ctx1, cancel1 := context.WithTimeout(context.Background(), timeout)
+	defer cancel1()
+	client, err := kivik.New(ctx1, "couch", connectionString)
```
Comment (drasko) [inline export/client/clients/couch-client.go]: `ctx`es are used only within `kivik.New()`, so you can reuse same `ctx, cancel` variables instead declaring new ones each time.

File: export/client/clients/couch-client.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+	defer cancel1()
+	rev, err := cc.Database.Rev(ctx1, reg.ID.Hex())
+	if err != nil {
+		return err
+	}
+
+	couchReg := ConvertToCouchReg(reg)
+	couchReg.Rev = rev
+
+	ctx2, cancel2 := context.WithTimeout(context.Background(), cc.Timeout)
```
Comment (drasko) [inline export/client/clients/couch-client.go]: No need for declaring new `ctx` var every time,you can reuse the old one.

File: export/client/clients/couch-client.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+		return export.Registration{}, err
+	}
+	return reg, err
+}
+
+func (cc *CouchClient) RegistrationByName(name string) (export.Registration, error) {
+	var reg export.Registration
+	var newReg CouchRegistration
+
+	findName := map[string]interface{}{"selector": map[string]interface{}{"name": map[string]interface{}{"$eq": name}}}
```
Comment (drasko) [inline export/client/clients/couch-client.go]: Indent this long map (break into several lines) to make it more readable.

File: export/client/clients/couch-client.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+func (cc *CouchClient) DeleteRegistrationById(id string) error {
+	_, err := cc.RegistrationById(id)
+	if err != nil {
+		return err
+	}
+	ctx1, cancel1 := context.WithTimeout(context.Background(), cc.Timeout)
+	defer cancel1()
+	rev, err := cc.Database.Rev(ctx1, id)
+
+	ctx2, cancel2 := context.WithTimeout(context.Background(), cc.Timeout)
```
Comment (drasko) [inline export/client/clients/couch-client.go]: You can reuse the same `ctx, cancel` vars.

File: export/client/const.go [hunk 6]
```diff
@@ -10,13 +10,13 @@ type ConfigurationStruct struct {
...
 	Port                 int
 	DBType               string
-	MongoURL             string
-	MongoUsername        string
-	MongoPassword        string
-	MongoDatabaseName    string
-	MongoPort            int
-	MongoConnectTimeout  int
-	MongoSocketTimeout   int
+	DBURL             string
```
Comment (drasko) [inline export/client/const.go]: Align indentation to other (existing) code.

**Comment by drasko** [comment]
@leecindy I like the idea having CouchDB offered as an extra choice. It's a nice DB with fantastic sync features, ideal for intermittently connected devices (do very good for EdgeX and IoT scenarios in general).

Please sync your branch to upstream and let's see how we can merge this.

**Comment by drasko** [comment]
No activity will prompt me (unfortunately) to close this PR and issue.

**Comment by drasko** [comment]
Closing due to inactivity. Feel free to reopen later.

**Pull request closed without merging**